mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -726,6 +887,76 @@ def set_tag(self, run_id, tag):
             self._check_run_is_active(run)
             session.merge(SqlTag(run_uuid=run_id, key=tag.key, value=tag.value))
 
+    def _set_tags(self, run_id, tags, attempt=None):
+        """
+        Set multiple tags on a run
+
+        :param run_id: String ID of the run
+        :param tags: List of RunTag instances to log
+        """
+        if not tags:
+            return
+
+        # use SessionManager instead of ManagedSessionManager to set
+        # isolation level for individual transaction.
+        # Here, we open a connection with SERIALIZABLE isolation level.
+        # it will be released when session.commit is called.
+        session = self.SessionMaker()
+        session.connection(execution_options={"isolation_level": "SERIALIZABLE"})
+
+        run = self._get_run(run_uuid=run_id, session=session)
+        self._check_run_is_active(run)
+
+        try:
+            current_tags = (
+                session.query(SqlTag)
+                .filter(SqlTag.run_uuid == run_id, SqlTag.key.in_([t.key for t in tags]))
+                .with_for_update()
+                .all()
+            )
+            current_tags = {t.key: t for t in current_tags}
+
+            new_tag_dict = {}
+            for tag in tags:
+                _validate_tag(tag.key, tag.value)
```

Comment:
Nit: Can we move tag validation to the top of the method so that we don't make any DB queries if tags are invalid?

Reply:
Yeah that's a good idea, done